packages/test/test-end-to-end-tests/src/test/mockSharedString.ts

```diff
@@ -0,0 +1,262 @@
+		id: string,
+		services: IChannelServices,
+		attributes: IChannelAttributes,
+	): Promise<MockInsertSharedString> {
+		const sharedObject = new MockInsertSharedString(id, runtime, attributes, "InsertString");
+		await sharedObject.load(services);
+		return sharedObject;
+	}
+	public create(document: IFluidDataStoreRuntime, id: string): MockInsertSharedString {
+		return new MockInsertSharedString(id, document, this.attributes, "InsertString");
+	}
+}
+
+// Note: other DDSes have called this variable snapshotFileName
+const headerBlobName = "header";
+interface ISnapshot {
+	inserts: ITrimmableInsert[];
+	baseText: string;
+	currentSequenceNumber: number;
+	minimumSequenceNumber: number;
+}
+
+interface IInsert {
+	text: string;
+	index: number;
+	referenceSequenceNumber: number;
+}
+
+export interface ITrimmableInsert extends IInsert {
+	sequenceNumber: number;
+}
+
+interface ILocalInsert extends IInsert {
+	localIndex: number;
+}
+
+interface IInsertOp extends IInsert {
+	type: "insert";
+	index: number;
+	text: string;
+}
+
+export interface IInsertStringEvent extends ISharedObjectEvents {
+	(event: "valueChanged", listener: (insert: ITrimmableInsert) => void);
+}
+
+function applyInsert(base: string, insert: IInsert): string {
+	return base.slice(0, insert.index) + insert.text + base.slice(insert.index);
+}
+
+// The goal of this dds is to test the minimum sequence number window
+// It is capable of inserting text only.
+// This DDS works very simply, it is insert only. All local edits are applied on top of the remote edits.
+// Trimming occurs when we process minimum sequence number greater than the reference sequence number of the beginning inserts.
+// The merge conflict resolution is rebase only.
+export class MockInsertSharedString extends SharedObject<IInsertStringEvent> {
+	static getFactory(): IChannelFactory {
+		return new MockInsertSharedStringFactory();
+	}
+
+	// In memory data that should be consistent across all clients at the same seq number
+	private _inserts: ITrimmableInsert[] = [];
+	private _baseText = "";
+	private _sequenceNumber = 0;
+	private _minimumSequenceNumber = 0;
+
+	// Local Memory - data that may not be consistent across clients as is impacted by local changes
+	private _localInsertIndex: number = 0;
+	private _localInserts: ILocalInsert[] = [];
+
+	// API Surface
+	// For testing purposes we expose inserts
+	public get inserts(): ITrimmableInsert[] {
+		return this._inserts;
+	}
+
+	// This would be a realistic API for exposing text
+	// All we are doing here is applying two sorted arrays by reference sequence number
+	public get text(): string {
+		let value = this._baseText;
+		let i = 0;
+		let j = 0;
+		while (i < this.inserts.length && j < this._localInserts.length) {
+			const remoteInsert = this.inserts[i];
+			const localInsert = this._localInserts[j];
+			if (remoteInsert.referenceSequenceNumber <= localInsert.referenceSequenceNumber) {
+				value = applyInsert(value, remoteInsert);
```

Comment:
This will not provide good outcome. The problem here - we interpret indexes for remote ops as if local ops applied before this op (that have appropriate reference sequence number). But in reality, all local ops will be acked after all known remote ops.
The right way to do so - apply all local changes at the end, but only after transforming their position, i.e. essentially implementing OT at this small scale.

Reply:
It's easy to demonstrate the problem. When processCore() processes local state, the result of this function should not change.